README.md
# Package ‘rapidphylo’

## Overview

`rapidphylo` is an R package for rapid estimation of tree-topology from large allele frequency data using Root Distances Method, under a Brownian Motion model. This is achieved by first computing the maximum likelihood estimates of the root distances, and then inferring the tree-topology from them. 

## Installation

Install the released version of this package from CRAN

```r
install.packages("rapidphylo")
```

Or install the development version of this package from GitHub

```r
remotes::install_github('ArindamRoyChoudhury/rapidphylo', upgrade="never")
```


## Example

```r
# A dataset "Human_Allele_Frequencies" is loaded with the package; 
# it has allele frequencies in 31,000 sites for 4 human populations and one outgroup human population.

# check data dimension
dim(Human_Allele_Frequencies)

# run RDM function
rd_tre <- RDM(Human_Allele_Frequencies, outgroup = "San", use = "pairwise.complete.obs")

# result visualization
plot(rd_tre, use.edge.length = FALSE, cex = 0.5)
```


NEWS.md
# rapidphylo 0.1.2

## Changes 

* Update the documentation of `RDM()`


# rapidphylo 0.1.1

## Changes 

* `RDM()` now can work with data from 5 populations.

* The example dataset "Human_Allele_Frequencies" now is a 5 × 44,000 matrix

# rapidphylo 0.1.0

## Changes

* Initial CRAN release


R/multiphylo.R
#' Compiling component subtrees into an object of class ‘multiPhylo’
#'
#' \code{multiphylo()} is used in function \code{topology_to_newick()}. It complies component subtrees and transform them into an object of class ‘multiPhylo’.
#'
#' @param base_tree A matrix, where each row represents a split, but without repetitions. In a given row, a k-way split is represented by assigning each set of taxa that is descendent of a split as an identifying number from 1 to k, and numbering the position corresponding to each taxa that is part of such a set by that identifying number. The other positions are set to zero.
#' @param label A character vector containing the names of the taxa.
#'
#' @import stats
#' @import phangorn
#' @import ape
#'
#' @return An object of class ‘multiPhylo’.
#' @noRd
#'
#' @examples NA
#'
multiphylo<-function(base_tree,label){

  outgroup<-label[length(label)]

  for (i in 1:dim(base_tree)[1]){
    ind3<-which(base_tree[i,]==0)
    ind4<-which(base_tree[i,]==1)
    ind5<-which(base_tree[i,]==2)
    if (length(ind3)==0){
      if (length(ind4)==0 & length(ind5)==0){
        grp1<-label[-length(label)]
        grp1<-paste("(",paste(grp1,collapse = ","),")",sep = "")
        grp<-paste("(",outgroup,",",grp1,");",sep="")
      }else{
        grp1<-label[ind4]
        if (length(ind4)>1){
          grp1<-paste("(",paste(grp1,collapse = ","),")",sep = "")
        }
        grp2<-label[ind5]
        if (length(ind5)>1){
          grp2<-paste("(",paste(grp2,collapse = ","),")",sep = "")
        }
        grp<-paste("(",outgroup,",(",grp1,",",grp2,"));",sep="")
      }
    }else{
      grp1<-label[ind4]
      if (length(ind4)>1){
        grp1<-paste("(",paste(grp1,collapse = ","),")",sep = "")
      }
      grp2<-label[ind5]
      if (length(ind5)>1){
        grp2<-paste("(",paste(grp2,collapse = ","),")",sep = "")
      }
      grp3<-label[ind3]
      if (length(ind3)>1){
        grp3<-paste("(",paste(grp3,collapse = ","),")",sep = "")
      }
      grp<-paste("(",outgroup,",(",grp1,",",grp2,",",grp3,"));",sep="")
    }
    tree<-read.tree(text = grp)
    mp<-list(tree)
    class(mp)<- "multiPhylo"
    if (i==1) mp2<-mp else mp2<-c(mp2,mp)
  }
  return(mp2)
}


R/RDM.R
#' Estimating tree-topology from allele frequency data
#'
#' \code{RDM()} estimates a tree-topology from allele frequencies.
#'
#' The input matrix is the observed values of the frequencies at tips \eqn{1, 2, ..., P, P+1}.
#' A logit transformation is performed on the allele frequency data, so that the observed values
#' are approximately normal. (The logit transformation of r refers to \eqn{\log\frac{r}{1-r}}.) The transformed matrix is converted into a data frame for further analyses.
#'
#' @param outgroup A variable that can be either the population name or a numerical row number of the outgroup data.
#' @param use Specify which part of data is used to compute the covariance matrix. The options are "\code{complete.obs}", "\code{pairwise.complete.obs}", "\code{everything}", "\code{all.obs}", and "\code{na.or.complete}". See \code{stats::cov} for more details.
#' @param mat_allele_freq A \eqn{(P+1) \times L} matrix containing the allele frequencies, where there are \eqn{P} taxa, plus one outgroup, and \eqn{L} loci.
#'
#' @return An estimated tree-topology in Newick format.
#' @export
#'
#' @references Peng J, Rajeevan H, Kubatko L, and RoyChoudhury A (2021) \emph{A fast likelihood approach for estimation of large phylogenies from continuous trait data}. Molecular Phylogenetics and Evolution 161 107142.
#'
#' @examples
#' # A dataset "Human_Allele_Frequencies" is loaded with the package;
#' # it has allele frequencies in 44,000 sites for
#' # 4 human populations and one outgroup human population.
#'
#' # check data dimension
#' dim(Human_Allele_Frequencies)
#'
#' # run RDM function
#' rd_tre <- RDM(Human_Allele_Frequencies, outgroup = "San", use = "pairwise.complete.obs")
#'
#' # result visualization
#' plot(rd_tre, use.edge.length = FALSE, cex = 0.5)
#'
RDM<-function(mat_allele_freq,outgroup,
              use=c("complete.obs","pairwise.complete.obs","everything","all.obs","na.or.complete")){

  mat_allele_freq[mat_allele_freq==1]<-0.99
  mat_allele_freq[mat_allele_freq==0]<-0.01
  trans_mat_allele_freq<-log(mat_allele_freq/(1-mat_allele_freq))
  trans_mat_allele_freq <- as.data.frame(trans_mat_allele_freq)

  use<-match.arg(use)
  names<-row.names(trans_mat_allele_freq)
  if (is.character(outgroup)){
    index<-which(names==outgroup)
  }else {
    index<-outgroup
  }
  trans_mat_allele_freq<-rbind(trans_mat_allele_freq[-index, ],trans_mat_allele_freq[index,])
  label<-row.names(trans_mat_allele_freq)
  array_zero_ID<-BMaf_to_zeroRDD(trans_mat_allele_freq,use=use)
  base_tree<-zeroRDD_to_splits(array_zero_ID)
  rd_tre<-topology_to_newick(base_tree,label=label)
  return(rd_tre)
}
